returnn/util/file_cache.py

```diff
@@ -523,23 +523,25 @@ def __init__(self, *, interval: float = 1.0, cache_base_dir: str):
         super().__init__(daemon=True)
         self.stop = Event()
         self.files = defaultdict(int)  # usage counter
+        self.locks: Dict[str, Lock] = {}  # filename -> lock
         self.interval = interval
         self.cache_base_dir = cache_base_dir
         self._is_started = False  # careful: `_started` is already a member of the base class
 
     def run(self):
         """thread main loop"""
         while True:
-            all_files = {}  # dict to have order deterministic
-            for filename in self.files.copy():  # copy dict under GIL to avoid modifications during iteration
-                all_files[filename] = True
-                all_files.update({k: True for k in _all_parent_dirs(filename, base_dir=self.cache_base_dir)})
-            for filename in all_files:
-                try:
-                    os.utime(filename, None)
-                except Exception as exc:
-                    print(f"FileCache: failed updating mtime of {filename}: {exc}")
-                    raise
+            # copy dicts under GIL to avoid modifications during iteration
+            locks = self.locks.copy()
```

Comment:
I wonder a bit here: When some other thread does `files_extend`, and creates a new lock (`self.locks[file] = Lock()`), there is a small race condition that two lock objects are created for the same `filename`. Here when we do `self.locks.copy`, it might copy the old lock instance, and not the one which survives in the end in `self.locks[filename]`. How much is this a problem? If this is a problem, how can we solve it?

Reply:
> Ah, can creating a lock release the GIL?

Maybe, but that's not what I mean. This code:
```python
            if file not in self.locks:
                self.locks[file] = Lock()
```
This code can release the GIL after the `if` check. If two threads concurrently do this, you can end up with this situation.

Reply:
Yes. The `Lock()` call will release the GIL, I believe.

Reply:
> I think I'll add a lock that guards any dict modifications (i.e. excluding any FS operations)

Yes that should solve it. The thread main loop actually doesn't need it, I think, as all the ops are safe (`self.locks.copy()`, and checking `filename not in self.files`).

Reply:
The code:
```python
            if file not in self.locks:
                self.locks[file] = Lock()
```
That are multiple statements. When the GIL gets released *after* `if file not in self.locks:`, but *before* `self.locks[file] = Lock()`, you have the problem. It doesn't even need to get to the `Lock()` call.

Reply:
>The thread main loop actually doesn't need it

I'll still add it just to avoid future accidental issues.

Reply:
```
In [6]: compile("if file not in self.locks: self.locks[file] = Lock()", "<source>", "exec")
Out[6]: <code object <module> at 0x1050eff50, file "<source>", line 1>

In [8]: dis.dis(_6)
  1           0 LOAD_NAME                0 (file)
              2 LOAD_NAME                1 (self)
              4 LOAD_ATTR                2 (locks)
              6 CONTAINS_OP              1
              8 POP_JUMP_IF_FALSE       22
             10 LOAD_NAME                3 (Lock)
             12 CALL_FUNCTION            0
             14 LOAD_NAME                1 (self)
             16 LOAD_ATTR                2 (locks)
             18 LOAD_NAME                0 (file)
             20 STORE_SUBSCR
        >>   22 LOAD_CONST               0 (None)
             24 RETURN_VALUE
```

Reply:
I mean for example after it executed the op in the bytecode offset 6, or offset 8, or 10, when it releases the GIL at that point. But it would be problematic at all points after 6 before 20.

Reply:
The GIL can be released at any time. (Otherwise pure computational multithreaded code could not work at all.)